### ISSUE TITLE
[Bug] Memory leak with {{#each}}

### ISSUE THREAD

**Issue by JackieChiles** [emberjs/ember.js#19033]
### 🐞 Describe the Bug
JS heap size grows substantially over time, indicating a memory leak, when updating an array rendered in an `{{#each}}` block in a component template. DOM nodes also increase seemingly without bound.

### 🔬 Minimal Reproduction
Click the "Run Tests" button in the Twiddle and observe JS heap size. Note that the Twiddle version of Ember is 3.18.1 as 3.19 doesn't appear to be supported yet, but I did try 3.17, 3.18.1, and 3.19 locally and they all show the same behavior. It also seems to affect both Glimmer and classic components.

https://ember-twiddle.com/5960c7f73dd7849064ebe43f22dbf4ed

### 😕 Actual Behavior
Over time, as an array bound to an `{{#each}}` block changes, JS heap size grows continuously. Manual garbage collection does not recover the allocated memory, nor does normal garbage collection when it occurs. The browser will eventually crash given large enough objects and enough time.

### 🤔 Expected Behavior
Expected heap size and number of DOM nodes to stop increasing.

### 🌍 Environment

- Ember: 3.17, 3.18.1, 3.19
- Node.js/npm: node 10.21.0, npm 6.14.4
- OS: Linux
- Browser: Chrome

### ➕ Additional Context
I did see https://github.com/emberjs/ember.js/issues/19020 but that ticket describes rendering components inside the loop. The example linked in this ticket does not render components. Also, the comments on that ticket indicate that the issue may have been resolved in Ember 3.19, but I can still reproduce with that version.

**Comment by rwjblue** [comment]
This was fixed in https://github.com/glimmerjs/glimmer-vm/pull/1113 (in glimmer-vm 0.54.1), and is already included in ember-source@3.21.0-beta.1+. I'll get that fix pulled into `release` branch for a future 3.20.1 patch release.

**Comment by rwjblue** [comment]
OK, pulled the changes back into `beta` (for release in `3.21.0-beta.2`) and tagged + kicked off publishing of 3.20.1.

https://github.com/emberjs/ember.js/releases/tag/v3.20.1

@JackieChiles - I'll update here once that is published to NPM, so that you can test against it...

**Comment by rwjblue** [comment]
OK, release is out. @JackieChiles mind confirming things are resolved (my local tests seem good)?

**Comment by JackieChiles** [comment]
I will check and let you know.

**Comment by steveszc** [comment]
FWIW I was seeing this memory leak occur in a production app immediately after our upgrade to 3.16.0.
Also, I am able to produce the memory leak in the reproduction twiddle on 3.16.5.

I can confirm that upgrading to 3.20.2 fixed the issue in our production app.

**Comment by pzuraq** [comment]
@steveszc that is strange, we wouldn't have expected the leak to show up until 3.17. I wonder if we had another leak that went undetected and was fixed by the VM upgrade..

**Comment by steveszc** [comment]
@pzuraq yeah that could be, the memory leak we were observing only really showed up on one of our most expensive routes. We were observing heap growth that was 2x pre3.16 and it seemed like GC was running less frequently and only major GC was actually freeing up memory. I have some memory profiles showing 3.16 vs 3.20 that I can post.

**Comment by steveszc** [comment]
_Before - Ember@3.16.10_
![image](https://user-images.githubusercontent.com/8826403/89911511-25ecde80-dbb7-11ea-88bb-6f3a588a57fb.png)


_After - Ember@3.20.3_
![image](https://user-images.githubusercontent.com/8826403/89910801-49fbf000-dbb6-11ea-824c-124f1026178e.png)

**Comment by NullVoxPopuli** [comment]
same demo as in OP, but in ember 7.3 (alpha):
https://limber.glimdown.com/edit?c=JYWwDg9gTgLgBAYQuCA7Apq%2BAzKy4DkAAgOYA2oI6UA9AMbKQZYEDcAUKJLHAN5xkAhjGpwAvnFz5i6EACNqNKAFdUZCBDBtOKHvxhRBdANboAJuMl4QhUhRBVaBo8eCoS2rtHj8AgpakbGXlFQShDAE9PXR84IxhgNADrW1kFWgg5ACt0OhhtdnQAD254M3RsQWUyeDohAGd6uABZWWgIgBV0epEoOGKRVDMmpBRmH3Y4KbgGVB6VPOgACggAdwwoABo4qBJ6gEo%2BSemT%2BuUwahX16m2wvf2OE5OYAAtgeoA6N2AEwTIACXQgjAAGVgAAvdBwAC8cFe7w%2B9VQwPqLwgMEBwLBkKWD2OTzhb0%2B2Dcf0xoIhUNh8M%2B31%2BAKBFMhj2mYnY%2BLgRGcJnMHLpwDJjOxVLgAAYOByuYYeWYOSTkQysZSYWKJQBINVSly8k6oZQgBAROrdFUARlF4vZJxI6HgLyFlIAIsBsNhcUcCVMoLblFBUISEfLBUrIXAALQB2moH4CxVM9AsqZsjk2u0OyHO13NABC7t4HJOsx6cHkKuawheHzwqjMSxpH3tIfQmewcBocHNFq7eM9cG9MF9-oABgASXjyCQ5oeJ8RW6apuCN%2BMABWodEwMAQL0E7nQeYL0yLtW3u5VdaJDfTzZdrfb54R-OD8bgAB8Xx3DgAqDsW-YfGAQAAYsARTmEsADMPaev2g5wKOvB0CeNpiAApNOHLJicWoyhydBGmQ3QAEqyII3zuCq9Z6ga%2BHdBKWHcqYsonHgqxNLCvi4nR0xIiiaIYle%2B69jBfpwKsbhmGsHwXFA2DQCAO7rh8VAgO0HzKPU5gAFIguSwozph862oRO4SSAlJmCCBhuCQgnQT6Inlq8VYmcguL-kBIFgQATFBSZzlMRDxIkqAcioqBdD09RLHhxoHB6npHnA0mJBYsLmjOhZoMWVGEWsbFwAArJaBIHlMNDtr4NTUMiCQAG5QsUgjgARcAQK23pgEIdDWYSUJhJE-RYMA3pkBEcDvON-qvFCDAOBu5gCG43QfHAHT2nA9RNVCyntAIQLGH23TVDAnylTMWXwCx%2BX1ldiYamqZ3lYg3rCH1cByIYQxwBgqw7JEB5quViVXSqHEPE8-lPFdVZtPVADy2S5CdSxXXiZ31jFBH1MR8lkSQKqY7RkMnLJfRLAR8DACq4rjXAAA8336rlrGsONADUbOHPmvbMXlUnqS8CM5HkSzczzBKEwAXDMNH1JsZ29uJxTS8A8viwSgjS-wtV-Mo6DS-WqbGUMyDmZZUDWe6Yhq%2BrJxyFrcA62QesGxeRsuWZkIWVZ7hWzbttTHQDtOy7kYfO7Jue%2BY5uW4c1sK56ZjB7r%2BthxHplmz7Nlx-7Aep9rKeuwi6em17Me%2B3HCdJr5TxsiVvZPQAqmAZivb1fZ5UleDSTAESPe2N2d7Ct3E9MLpwNFstwAAfGKXNVwIr1QPecuSKoeTBRP88B9MN2qBFyOE00EamjXvbW0l1ApWfrL9GQGnxeL9ZBnGwoURePFgKi6K6ZSnFnXXVkc46YiGaq9aeHIQGCDkARCBvYQGNjMHA9WICoDIJ3gg6edMaCvHQQHTBC94EvHAtPckcBhTYOIXg222DcEL1oWgs6tDEF4KYTAOQEAzARGoZ6VBPDxYgKQYQ3hVCACS0Z6SUJIcIqYtChE71kTACwdRBCNGhAAIiXGGeolJ1H8IDrwXg9ZHyv0pGIQBGCcHyJQTgxhJx6EGH0fApRTiUFUOAgqKRriCRyO8U8QRMwGj1A0VonRkI9EyKeIY5%2BpJTGQnMZE3x9DbGsJoY4xJLjIknDplQlsXjElWL8dkpRgTVHBM0YybRuiimemiReJcwoWwJIUfTQpyT0n2LSXYhRgiak%2BKoTmOAeSaBUIKZklpASVFqIqcCKp4S%2BlRKMfUq8LYczNJ6W0mhKSOQOO6Rg8ZLTpg5JIShRASF0D5ImZsnpJSpnlNCdUrJ0w6kIgaZSVcUB1xYC3DuZCFj8HXIEds%2BBOCOFcLwbQ6BsCdnZLkMoGAAF-SGKSOouowATDqIntgKaF4woHyaJRfUhpYr7HMdPQiqhVrdBOtguFCK0B4OJVjQ6uNozuC1kshER8cakTZSQdZUwOSGIAMRAkQmHEGqjXwsRfAK-xZhgC1T8XTeoYAdzT0MSxD4ggPgh3QOY7Bqr1X0KNagDVvAtVyF1SnA1NBTXKvtZqySdBrXO31WIQ1aqzUmq9earVZhXV61tfan16qnWrA%2BOgQN7rPXGpBQqpVQreA0DFS8OVkCegRGhb2JRl5ZlhKhGLcWsksDSxUqgCARr1wZU9CWmAczU5eQACxgCKDWgkIgij1r%2BMAEgqBpYWxIC8GA7bZzZLtb3bNrTQGdXAewNkQA&format=gjs

**Comment by NullVoxPopuli** [comment]
No memory leak is present.
